File: JavaScript/07_ReverseInteger.js
/**
 * @param {number} x
 * @return {number}
 */
const reverse = function(x) {
    let negative = false;
    let revNum = 0;

    if (x < 0) {
        negative = true;
        x = Math.abs(x);
    }
    
    while (x > 0) {
        let lastDigit = x % 10;
        revNum = revNum * 10 + lastDigit;
        x = Math.floor(x / 10); 
    }

    if (revNum > (2**31 - 1)) {
        return 0;
    }
    
    return negative ? -revNum : revNum;
};

const testCases = [
    {
        x: -1234,
        expected: -4321
    },
    {
        x: 1534236469,
        expected: 0
    },
    {
        x: 123,
        expected: 321
    },
    {
        x: 120,
        expected: 21
    }
]

testCases.forEach((testCase, i) => {
    const output = reverse(testCase.x);
    if (JSON.stringify(output) === JSON.stringify(testCase.expected)) {
        console.log(`Test Case ${i}: Passed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    } else {
        console.log(`Test Case ${i}: Failed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    }
});

File: C/07_ReverseInteger.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int reverse(int x) {
    int revNum = 0;

    while (x != 0) {
        int lastDigit = x % 10;

        if (revNum > INT_MAX / 10 || (revNum == INT_MAX / 10 && lastDigit > 7)) {
            return 0; 
        }
        if (revNum < INT_MIN / 10 || (revNum == INT_MIN / 10 && lastDigit < -8)) {
            return 0; 
        }

        revNum = revNum * 10 + lastDigit;
        x /= 10;
    }

    return revNum;
}

void runTests() {
    printf("Running Reverse Integer Tests...\n");

    struct TestCase {
        int input;
        int expected;
    } testCases[] = {
        {-1234, -4321},
        {1534236469, 0},
        {123, 321},
        {120, 21}
    };

    int numTests = sizeof(testCases) / sizeof(testCases[0]);

    for (int i = 0; i < numTests; i++) {
        int result = reverse(testCases[i].input);
        if (result == testCases[i].expected) {
            printf("Test Case %d Passed: Input: %d, Expected: %d, Result: %d\n",
                i + 1, testCases[i].input, testCases[i].expected, result);
        } else {
            printf("Test Case %d Failed: Input: %d, Expected: %d. Result: %d\n",
                i + 1, testCases[i].input, testCases[i].expected, result);
        }
    }
}

int main() {
    runTests();
    return 0;
}